File: MovieCatalogue_Java_Submission_1/app/src/main/java/id/apwdevs/moviecatalogue/presenter/MainListMoviePresenter.java
package id.apwdevs.moviecatalogue.presenter;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;

import java.util.ArrayList;
import java.util.List;

import id.apwdevs.moviecatalogue.R;
import id.apwdevs.moviecatalogue.model.ShortListModel;
import id.apwdevs.moviecatalogue.view.MainListMovieView;

public class MainListMoviePresenter {
    private Context mContext;
    private MainListMovieView view;

    public MainListMoviePresenter(Context mContext, MainListMovieView view) {
        this.mContext = mContext;
        this.view = view;
    }

    public void prepareAll() {
        view.onLoadData();

        Resources resources = mContext.getResources();
        String[] arrMovies = resources.getStringArray(R.array.movie_name_list);
        String[] arrOverview = resources.getStringArray(R.array.movie_overview_str_list);
        String[] arrReleased = resources.getStringArray(R.array.movie_released_time_list);
        TypedArray arrMoviePoster = resources.obtainTypedArray(R.array.movie_drawable_list);
        List<ShortListModel> listResult = new ArrayList<>();
        for (int x = 0; x < arrMovies.length; x++) {
            ShortListModel shortListModel = new ShortListModel();
            shortListModel.setTitleMovie(arrMovies[x]);
            shortListModel.setOverview(arrOverview[x]);
            shortListModel.setReleaseDate(arrReleased[x]);
            shortListModel.setPhotoRes(arrMoviePoster.getResourceId(x, -1));
            listResult.add(shortListModel);
        }
        arrMoviePoster.recycle();
        view.onLoadFinished(listResult);
        System.gc();
    }
}

File: MovieCatalogue_Kotlin_Submission_2/app/src/main/java/id/apwdevs/moTvCatalogue/presenter/MainListMoviePresenter.kt
package id.apwdevs.moTvCatalogue.presenter

import android.content.Context
import android.graphics.Point
import id.apwdevs.moTvCatalogue.R
import id.apwdevs.moTvCatalogue.model.ShortListModel
import id.apwdevs.moTvCatalogue.view.MainListTvOrMovieView
import java.util.*

class MainListMoviePresenter(
    private val mContext: Context,
    private val view: MainListTvOrMovieView,
    private val windowSize: Point
) {

    fun prepareAll() {
        view.onLoadData()

        val resources = mContext.resources
        val arrMovies = resources.getStringArray(R.array.movie_name_list)
        val arrOverview = resources.getStringArray(R.array.movie_overview_str_list)
        val arrReleased = resources.getStringArray(R.array.movie_released_time_list)
        val arrMoviePoster = resources.obtainTypedArray(R.array.movie_drawable_list)
        val listResult = ArrayList<ShortListModel>()
        for (x in arrMovies.indices) {
            listResult.add(
                ShortListModel(
                    arrMoviePoster.getResourceId(x, -1),
                    arrOverview[x],
                    arrReleased[x],
                    arrMovies[x]
                )
            )
        }
        arrMoviePoster.recycle()
        view.onLoadFinished(listResult, calculateMaxColumn())
    }

    private fun calculateMaxColumn(): Int {
        val resources = mContext.resources
        val vPagerMarginStart = resources.getDimension(R.dimen.viewpager_user_marginstart)
        val vPagerMarginEnd = resources.getDimension(R.dimen.viewpager_user_marginstart)
        val cardViewPadding = resources.getDimension(R.dimen.cardview_grid_content_padding)
        val imageViewWidth = resources.getDimension(R.dimen.item_poster_width)

        var maxColumn = 1
        while (true) {
            val contentMeasuredWidth = imageViewWidth + 2 * cardViewPadding
            val spaceFreeForContent =
                windowSize.x.toFloat() - (vPagerMarginStart + vPagerMarginEnd) - maxColumn * contentMeasuredWidth
            if (spaceFreeForContent > contentMeasuredWidth) {
                maxColumn++
            } else {
                break
            }
        }
        return maxColumn
    }
}